### PR TITLE
Only use lives (and not aliases) to create post parfor live set.

### LINKED ISSUE #4690: numba 0.46, useless(?) code in parfor init block
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [ ] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

We are porting hpat to numba 0.46 and found discrepancies in parfor pass output.

pre parfor pass input IR is identical.

In previous version init block after parfor pass was empty.

In 0.46 init block contains seemingly useless code.
```
init block:
    arr.83 = $arr.83.139                     ['$arr.83.139', 'arr.83']
    del arr.83                               []
```

We didn't manage to reduce code to pure numba yet but here are IR before preparfor pass  and after parfor pass.

Any ideas why this is happens?

prev numba
```
=========== before preparfor ===========
label 0:
    N = arg(0, name=N)                       ['N']
    $0.1 = global(np: <module 'numpy' from '/home/ibutygin/miniconda3/envs/HPAT/lib/python3.7/site-packages/numpy/__init__.py'>) ['$0.1']
    $0.2 = getattr(value=$0.1, attr=arange)  ['$0.1', '$0.2']
    $0.4 = call $0.2(N, func=$0.2, args=[Var(N, /home/ibutygin/src/hpat/hpat/hpat/tests/test_basic.py (303))], kws=(), vararg=None) ['$0.2', '$0.4', 'N']
    A = $0.4                                 ['$0.4', 'A']
    _dist_arr.36 = A                         ['A', '_dist_arr.36']
    $0.1.37 = global(hpat: <module 'hpat' from '/home/ibutygin/src/hpat/hpat/hpat/__init__.py'>) ['$0.1.37']
    $0.2.38 = getattr(value=$0.1.37, attr=distributed_api) ['$0.1.37', '$0.2.38']
    $0.3.39 = getattr(value=$0.2.38, attr=dist_return) ['$0.2.38', '$0.3.39']
    $0.5.40 = call $0.3.39(_dist_arr.36, func=$0.3.39, args=[Var(_dist_arr.36, /home/ibutygin/src/hpat/hpat/hpat/hiframes/hiframes_untyped.py (1735))], kws=(), vararg=None) ['$0.3.39', '$0.5.40', '_dist_arr.36']
    _d_arr.41 = $0.5.40                      ['$0.5.40', '_d_arr.41']
    distributed_return.44 = cast(value=_d_arr.41) ['_d_arr.41', 'distributed_return.44']
    return distributed_return.44             ['distributed_return.44']
========================================
============= after parfor =============
label 0:
    N = arg(0, name=N)                       ['N']
    $start.54.115 = const(int, 0)            ['$start.54.115']
    $step.56.116 = const(int, 1)             ['$step.56.116']
    $dtype.57.117 = const(dtype, int64)      ['$dtype.57.117']
    $0.5.62 = global(math: <module 'math' from '/home/ibutygin/miniconda3/envs/HPAT/lib/python3.7/lib-dynload/math.cpython-37m-x86_64-linux-gnu.so'>) ['$0.5.62']
    $0.6.63 = getattr(value=$0.5.62, attr=ceil) ['$0.5.62', '$0.6.63']
    del $0.5.62                              []
    $0.9.66 = N - $start.54.115              ['$0.9.66', '$start.54.115', 'N']
    del N                                    []
    $0.11.68 = $0.9.66 / $step.56.116        ['$0.11.68', '$0.9.66', '$step.56.116']
    del $0.9.66                              []
    $nitems_r.70.118 = call $0.6.63($0.11.68, func=$0.6.63, args=[Var($0.11.68, /home/ibutygin/miniconda3/envs/HPAT/lib/python3.7/site-packages/numba/parfor.py (354))], kws=(), vararg=None) ['$0.11.68', '$0.6.63', '$nitems_r.70.118']
    del $0.6.63                              []
    del $0.11.68                             []
    $0.13.71 = global(int: <class 'int'>)    ['$0.13.71']
    $0.14.72 = global(max: <built-in function max>) ['$0.14.72']
    $const0.16.74 = const(int, 0)            ['$const0.16.74']
    $0.17.75 = call $0.14.72($nitems_r.70.118, $const0.16.74, func=$0.14.72, args=[Var($nitems_r.70.118, /home/ibutygin/miniconda3/envs/HPAT/lib/python3.7/site-packages/numba/parfor.py (354)), Var($const0.16.74, /home/ibutygin/miniconda3/envs/HPAT/lib/python3.7/site-packages/numba/parfor.py (355))], kws=(), vararg=None) ['$0.14.72', '$0.17.75', '$const0.16.74', '$nitems_r.70.118']
    del $nitems_r.70.118                     []
    del $const0.16.74                        []
    del $0.14.72                             []
    $nitems.77.119 = call $0.13.71($0.17.75, func=$0.13.71, args=[Var($0.17.75, /home/ibutygin/miniconda3/envs/HPAT/lib/python3.7/site-packages/numba/parfor.py (355))], kws=(), vararg=None) ['$0.13.71', '$0.17.75', '$nitems.77.119']
    del $0.17.75                             []
    del $0.13.71                             []
    $0.19.78 = global(np: <module 'numpy' from '/home/ibutygin/miniconda3/envs/HPAT/lib/python3.7/site-packages/numpy/__init__.py'>) ['$0.19.78']
    $0.20.79 = getattr(value=$0.19.78, attr=empty) ['$0.19.78', '$0.20.79']
    del $0.19.78                             []
    $arr.83.120 = call $0.20.79($nitems.77.119, $dtype.57.117, func=$0.20.79, args=[Var($nitems.77.119, /home/ibutygin/miniconda3/envs/HPAT/lib/python3.7/site-packages/numba/parfor.py (355)), Var($dtype.57.117, /home/ibutygin/miniconda3/envs/HPAT/lib/python3.7/site-packages/numba/parfor.py (332))], kws=(), vararg=None) ['$0.20.79', '$arr.83.120', '$dtype.57.117', '$nitems.77.119']
    del $dtype.57.117                        []
    del $0.20.79                             []
    $0.1.37 = global(hpat: <module 'hpat' from '/home/ibutygin/src/hpat/hpat/hpat/__init__.py'>) ['$0.1.37']
    $0.2.38 = getattr(value=$0.1.37, attr=distributed_api) ['$0.1.37', '$0.2.38']
    del $0.1.37                              []
    $0.3.39 = getattr(value=$0.2.38, attr=dist_return) ['$0.2.38', '$0.3.39']
    del $0.2.38                              []
---begin parfor 0---
index_var =  parfor_index.112
LoopNest(index_variable = parfor_index.112, range = (0, $nitems.77.119, 1))
init block:
label 78:
    i.99 = parfor_index.112                  ['i.99', 'parfor_index.112']
    $74.5.103 = i.99 * $step.56.116          ['$74.5.103', '$step.56.116', 'i.99']
    $74.6.104 = $start.54.115 + $74.5.103    ['$74.5.103', '$74.6.104', '$start.54.115']
    $arr.83.120[parfor_index.112] = $74.6.104 ['$74.6.104', '$arr.83.120', 'parfor_index.112']
----end parfor 0----
    del $step.56.116                         []
    del $start.54.115                        []
    $_d_arr.41.121 = call $0.3.39($arr.83.120, func=$0.3.39, args=[Var($arr.83.120, /home/ibutygin/miniconda3/envs/HPAT/lib/python3.7/site-packages/numba/parfor.py (356))], kws=(), vararg=None) ['$0.3.39', '$_d_arr.41.121', '$arr.83.120']
    del $arr.83.120                          []
    del $0.3.39                              []
    distributed_return.44 = cast(value=$_d_arr.41.121) ['$_d_arr.41.121', 'distributed_return.44']
    del $_d_arr.41.121                       []
    return distributed_return.44             ['distributed_return.44']
========================================
```

numba 0.46
```
=========== before preparfor ===========
label 0:
    N = arg(0, name=N)                       ['N']
    $0.1 = global(np: <module 'numpy' from '/home/ibutygin/miniconda3/envs/NUMBA_CUSTOM/lib/python3.7/site-packages/numpy/__init__.py'>) ['$0.1']
    $0.2 = getattr(value=$0.1, attr=arange)  ['$0.1', '$0.2']
    $0.4 = call $0.2(N, func=$0.2, args=[Var(N, /home/ibutygin/src/numba/hpat/hpat/tests/test_basic.py (303))], kws=(), vararg=None) ['$0.2', '$0.4', 'N']
    A = $0.4                                 ['$0.4', 'A']
    _dist_arr.36 = A                         ['A', '_dist_arr.36']
    $0.1.37 = global(hpat: <module 'hpat' from '/home/ibutygin/src/numba/hpat/hpat/__init__.py'>) ['$0.1.37']
    $0.2.38 = getattr(value=$0.1.37, attr=distributed_api) ['$0.1.37', '$0.2.38']
    $0.3.39 = getattr(value=$0.2.38, attr=dist_return) ['$0.2.38', '$0.3.39']
    $0.5.40 = call $0.3.39(_dist_arr.36, func=$0.3.39, args=[Var(_dist_arr.36, /home/ibutygin/src/numba/hpat/hpat/hiframes/hiframes_untyped.py (1753))], kws=(), vararg=None) ['$0.3.39', '$0.5.40', '_dist_arr.36']
    _d_arr.41 = $0.5.40                      ['$0.5.40', '_d_arr.41']
    distributed_return.44 = cast(value=_d_arr.41) ['_d_arr.41', 'distributed_return.44']
    return distributed_return.44             ['distributed_return.44']
========================================
============= after parfor =============
label 0:
    N = arg(0, name=N)                       ['N']
    $start.54.115 = const(int, 0)            ['$start.54.115']
    $step.56.116 = const(int, 1)             ['$step.56.116']
    $dtype.57.117 = const(dtype, int64)      ['$dtype.57.117']
    $0.5.62 = global(math: <module 'math' from '/home/ibutygin/miniconda3/envs/NUMBA_CUSTOM/lib/python3.7/lib-dynload/math.cpython-37m-x86_64-linux-gnu.so'>) ['$0.5.62']
    $0.6.63 = getattr(value=$0.5.62, attr=ceil) ['$0.5.62', '$0.6.63']
    del $0.5.62                              []
    $0.9.66 = N - $start.54.115              ['$0.9.66', '$start.54.115', 'N']
    del N                                    []
    $0.11.68 = $0.9.66 / $step.56.116        ['$0.11.68', '$0.9.66', '$step.56.116']
    del $0.9.66                              []
    $nitems_r.70.118 = call $0.6.63($0.11.68, func=$0.6.63, args=[Var($0.11.68, /home/ibutygin/src/numba/numba/numba/parfor.py (354))], kws=(), vararg=None) ['$0.11.68', '$0.6.63', '$nitems_r.70.118']
    del $0.6.63                              []
    del $0.11.68                             []
    $0.13.71 = global(int: <class 'int'>)    ['$0.13.71']
    $0.14.72 = global(max: <built-in function max>) ['$0.14.72']
    $const0.16.74 = const(int, 0)            ['$const0.16.74']
    $0.17.75 = call $0.14.72($nitems_r.70.118, $const0.16.74, func=$0.14.72, args=[Var($nitems_r.70.118, /home/ibutygin/src/numba/numba/numba/parfor.py (354)), Var($const0.16.74, /home/ibutygin/src/numba/numba/numba/parfor.py (355))], kws=(), vararg=None) ['$0.14.72', '$0.17.75', '$const0.16.74', '$nitems_r.70.118']
    del $nitems_r.70.118                     []
    del $const0.16.74                        []
    del $0.14.72                             []
    $nitems.77.119 = call $0.13.71($0.17.75, func=$0.13.71, args=[Var($0.17.75, /home/ibutygin/src/numba/numba/numba/parfor.py (355))], kws=(), vararg=None) ['$0.13.71', '$0.17.75', '$nitems.77.119']
    del $0.17.75                             []
    del $0.13.71                             []
    $0.19.78 = global(np: <module 'numpy' from '/home/ibutygin/miniconda3/envs/NUMBA_CUSTOM/lib/python3.7/site-packages/numpy/__init__.py'>) ['$0.19.78']
    $0.20.79 = getattr(value=$0.19.78, attr=empty) ['$0.19.78', '$0.20.79']
    del $0.19.78                             []
    $arr.83.139 = call $0.20.79($nitems.77.119, $dtype.57.117, func=$0.20.79, args=[Var($nitems.77.119, /home/ibutygin/src/numba/numba/numba/parfor.py (355)), Var($dtype.57.117, /home/ibutygin/src/numba/numba/numba/parfor.py (332))], kws=(), vararg=None) ['$0.20.79', '$arr.83.139', '$dtype.57.117', '$nitems.77.119']
    del $dtype.57.117                        []
    del $0.20.79                             []
    $0.1.37 = global(hpat: <module 'hpat' from '/home/ibutygin/src/numba/hpat/hpat/__init__.py'>) ['$0.1.37']
    $0.2.38 = getattr(value=$0.1.37, attr=distributed_api) ['$0.1.37', '$0.2.38']
    del $0.1.37                              []
    $0.3.39 = getattr(value=$0.2.38, attr=dist_return) ['$0.2.38', '$0.3.39']
    del $0.2.38                              []
---begin parfor 0---
index_var =  parfor_index.112
LoopNest(index_variable = parfor_index.112, range = (0, $nitems.77.119, 1))
init block:
    arr.83 = $arr.83.139                     ['$arr.83.139', 'arr.83']
    del arr.83                               []
label 78:
    i.99 = parfor_index.112                  ['i.99', 'parfor_index.112']
    $74.5.103 = i.99 * $step.56.116          ['$74.5.103', '$step.56.116', 'i.99']
    $74.6.104 = $start.54.115 + $74.5.103    ['$74.5.103', '$74.6.104', '$start.54.115']
    $arr.83.139[parfor_index.112] = $74.6.104 ['$74.6.104', '$arr.83.139', 'parfor_index.112']
----end parfor 0----
    del $step.56.116                         []
    del $start.54.115                        []
    $_d_arr.41.121 = call $0.3.39($arr.83.139, func=$0.3.39, args=[Var($arr.83.139, /home/ibutygin/src/numba/numba/numba/parfor.py (356))], kws=(), vararg=None) ['$0.3.39', '$_d_arr.41.121', '$arr.83.139']
    del $arr.83.139                          []
    del $0.3.39                              []
    distributed_return.44 = cast(value=$_d_arr.41.121) ['$_d_arr.41.121', 'distributed_return.44']
    del $_d_arr.41.121                       []
    return distributed_return.44             ['distributed_return.44']
========================================
```

### PR DESCRIPTION
**Pull request by DrTodd13**
Resolves #4690

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the patch. @Hardcode84 please could you confirm that the patch herein fixes the issue presented https://github.com/numba/numba/issues/4690, thanks.

**Comment by Hardcode84** [comment]
Hmm, our HPAT PR is based on numba release.46 branch and seems current numba master diverged enough from it and our code got broken again (unrelated to parfor). I checked initial version of **IntelLabs:issue4690** branch and it worked fine but I cannot check current version right now. We will update to numba master eventually after merging 0.46 version (with disabled tests broken due to parfor) but it will take some time.

**Comment by Hardcode84** [comment]
Hi @DrTodd13 @stuartarchibald we checked this branch and it fixes issue, thanks.

**Comment by stuartarchibald** [comment]
@Hardcode84 thanks for confirming.

**Comment by Hardcode84** [comment]
@DrTodd13 @stuartarchibald any update on this?

**Comment by stuartarchibald** [comment]
@DrTodd13 please could you resolve the conflicts on this? Thanks.

**Comment by stuartarchibald** [comment]
@ehsantn I would like to review this with view to merge as it fixes a known problem. IIRC at a previous core developer meeting you raised some concerns around this approach. Please could you enumerate these concerns and propose alternative solutions? Thanks.

**Comment by ehsantn** [comment]
This PR essentially says ignore aliases of live variables during recursive dead code elimination of parfors. This sounds to me like introducing an obvious correctness issue to "improve" optimization results.

The underlying problem is that copy propagation doesn't handle parfor init blocks properly. A quick workaround would to scan the init block and remove dead assignments.

**Comment by DrTodd13** [comment]
> This PR essentially says ignore aliases of live variables during recursive dead code elimination of parfors. This sounds to me like introducing an obvious correctness issue to "improve" optimization results.
> 
> The underlying problem is that copy propagation doesn't handle parfor init blocks properly. A quick workaround would to scan the init block and remove dead assignments.

@ehsantn I don't understand your suggestion.  Recursive DCE already scans the init block to remove dead assignments.  I believe that the code in this PR still maintains setitem's in the parfor (init block or otherwise) if those setitems are for arrays which alias and one of them is live after the parfor.  If a setitem on an alias is present then the alias assignment itself will be maintained.  (See the use of alias_map there.)  So, if we have "dummy=arr" in the init_block and "arr[index] = something" in the loop body and arr survives the parfor then this doesn't imply that the "dummy" var needs to survive.

**Comment by ehsantn** [comment]
@DrTodd13 Yes, setitem handling and other parts of parfor dead code elimination still use alias information with changes of this PR. However, parfor bodies can have almost any kind of code, not just setitems. 

Yes, the "dummy" var shouldn't survive. Aliases are not used in removing assignments so it should have been removed.

**Comment by ehsantn** [comment]
@DrTodd13 I just had a chance to dip deeper using the test case in this PR. I think this change is actually safe, since aliases are not considered only in [_add_liveness_return_block](https://github.com/numba/numba/blob/ef119bcd1733ff49d71bdf2da8a66e91bb704f83/numba/parfor.py#L3734) inside [remove_dead_parfor_recursive](https://github.com/numba/numba/blob/ef119bcd1733ff49d71bdf2da8a66e91bb704f83/numba/parfor.py#L3725). However, `alias_map` of the whole function is reused for recursive dead code removal and not recomputed. This makes [assignment handling](https://github.com/numba/numba/blob/ef119bcd1733ff49d71bdf2da8a66e91bb704f83/numba/ir_utils.py#L594) of `remove_dead_block` work properly since it allows distinguishing between live variables and their aliases. 

Does this make sense? If so, this needs to be documented in the code due to complexity.

**Comment by DrTodd13** [comment]
> @DrTodd13 I just had a chance to dip deeper using the test case in this PR. I think this change is actually safe, since aliases are not considered only in [_add_liveness_return_block](https://github.com/numba/numba/blob/ef119bcd1733ff49d71bdf2da8a66e91bb704f83/numba/parfor.py#L3734) inside [remove_dead_parfor_recursive](https://github.com/numba/numba/blob/ef119bcd1733ff49d71bdf2da8a66e91bb704f83/numba/parfor.py#L3725). However, `alias_map` of the whole function is reused for recursive dead code removal and not recomputed. This makes [assignment handling](https://github.com/numba/numba/blob/ef119bcd1733ff49d71bdf2da8a66e91bb704f83/numba/ir_utils.py#L594) of `remove_dead_block` work properly since it allows distinguishing between live variables and their aliases.
> 
> Does this make sense? If so, this needs to be documented in the code due to complexity.

@ehsantn Thanks for your thoughtful review and comments.  I've added some comments to the code.  If only that were the only complex spot that was still not documented.  I stole a little bit of your text for the comment.  Hope you don't mind.

**Comment by ehsantn** [comment]
Thank you, @DrTodd13. Looks great. I'm glad we finally figured this out!

File: numba/parfor.py [hunk 1]
```diff
@@ -3731,6 +3737,17 @@ def remove_dead_parfor_recursive(parfor, lives, arg_aliases, alias_map,
...
     last_label = max(blocks.keys())
 
+    """
+      Previously, this statement used lives_n_aliases.  That had the effect of
+      keeping variables in the init_block alive if they aliased an array that
+      was later written to.  By using just lives to indicate which variables
+      names are live at exit of the parfor but then using alias_map for the
+      actual recursive dead code removal, we keep any writes to aliased arrays
+      alive but also allow aliasing assignments (i.e., a = b) to be eliminated
+      so long as 'a' is not written to through the variable 'a' later on.
```
Comment (stuartarchibald) [inline numba/parfor.py]:
```suggestion
      so long as 'a' is not written to through the variable 'b' later on.
```
do you meant that ^ ?

**Comment by stuartarchibald** [comment]
@DrTodd13 @ehsantn thanks very much to you both for working through this difficult issue, it's much appreciated. It seems we've reached an agreement, the code looks good and the docs/prose are very helpful. This should make 0.48.

File: numba/parfor.py (hunk 1 above)
Comment (DrTodd13) [reply numba/parfor.py]: I think it is the reverse.  I've pushed a fix. Thanks for the catch @stuartarchibald .

Comment (stuartarchibald) [reply numba/parfor.py]: Great, thanks!

**Pull request merged**